src/mscorlib/src/System/IO/File.cs

```diff
@@ -821,5 +823,323 @@ internal static int FillAttributeInfo(String path, ref Win32Native.WIN32_FILE_AT
+            if (path.Length == 0)
+                throw new ArgumentException(SR.Argument_EmptyPath, nameof(path));
+
+            Contract.EndContractBlock();
+
+            return cancellationToken.IsCancellationRequested
+                ? Task.FromCanceled<string[]>(cancellationToken)
+                : InternalReadAllLinesAsync(path, encoding, cancellationToken);
+        }
+
+        private static async Task<string[]> InternalReadAllLinesAsync(
+            String path, Encoding encoding, CancellationToken cancel)
+        {
+            Contract.Requires(!string.IsNullOrEmpty(path));
+            Contract.Requires(encoding != null);
+
+            using (StreamReader sr = AsyncStreamReader(path, encoding))
+            {
+                cancel.ThrowIfCancellationRequested();
+                string line;
+                List<string> lines = new List<string>();
+                while ((line = await sr.ReadLineAsync().ConfigureAwait(false)) != null)
+                {
+                    cancel.ThrowIfCancellationRequested();
+                    lines.Add(line);
+                }
+
+                cancel.ThrowIfCancellationRequested();
+                return lines.ToArray();
+            }
+        }
+
+        public static Task WriteAllLinesAsync(
+            string path, IEnumerable<string> contents, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return WriteAllLinesAsync(path, contents, StreamWriter.UTF8NoBOM, cancellationToken);
+        }
+
+        public static Task WriteAllLinesAsync(
+            string path, IEnumerable<string> contents, Encoding encoding,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+            if (contents == null)
+                throw new ArgumentNullException(nameof(contents));
+            if (encoding == null)
+                throw new ArgumentNullException(nameof(encoding));
+            if (path.Length == 0)
+                throw new ArgumentException(SR.Argument_EmptyPath, nameof(path));
+
+            Contract.EndContractBlock();
+
+            return cancellationToken.IsCancellationRequested
+                ? Task.FromCanceled(cancellationToken)
+                : InternalWriteAllLinesAsync(AsyncStreamWriter(path, false, encoding), contents, cancellationToken);
+        }
+
+        private static async Task InternalWriteAllLinesAsync(TextWriter writer, IEnumerable<String> contents, CancellationToken cancel)
+        {
+            Contract.Requires(writer != null);
+            Contract.Requires(contents != null);
+
+            using (writer)
+            {
+                foreach (String line in contents)
+                {
+                    cancel.ThrowIfCancellationRequested();
+                    await (line == null ? writer.WriteLineAsync() : writer.WriteLineAsync(line)).ConfigureAwait(false);
```

Comment:
Having a null-check where there is none in the non-async version is a workaround for #8661